### PR TITLE
[IBCDPE-413]/test load.py

### PR DESCRIPTION
**Pull request by BWMac**
This PR implements tests for `test_load.py`, improves docstrings for `load.py`, and makes some proposed changes to functions in `load.py` for bugs and edge cases that I came across while writing the tests. I will go through and add more detailed comments in some of these areas and I would love to get some feedback on those changes to make sure that they are useful and appropriate.

### CODE REVIEW

File: agoradatatools/etl/load.py [hunk 1]
```diff
@@ -1,9 +1,12 @@
...
+from typing import Union
+
 import numpy as np
+import pandas as pd
+from synapseclient import Activity, File, Synapse
+
+from . import utils
 
 
 class NumpyEncoder(json.JSONEncoder):
```
Comment (BWMac) [inline agoradatatools/etl/load.py:12]: I did not do any testing for the `NumpyEncoder` because I noticed that there was a specific ticket for this further down in my backlog.

File: agoradatatools/etl/load.py [hunk 2]
```diff
@@ -30,18 +36,25 @@ def create_temp_location(staging_path: str):
...
-    Deletes the default temporary location
+    """Deletes the default temporary location
+
+    Args:
+        staging_path (str): path to temporary directory to be deleted
     """
     rmdir(staging_path)
 
 
 def remove_non_values(d: dict) -> dict:
```
Comment (BWMac) [inline agoradatatools/etl/load.py:47]: This is the area where I made the biggest changes in this file. The issue that I noticed and was attempting to provide a solution for is that the `remove_non_values` function was not actually removing keys from nested dictionaries whose values are `None`. The test case that I used was `{"d": {"e": "f", "g": None}}`, and the exact same dictionary was being returned back by the original version of the function when I was expecting `{"d": {"e": "f"}` where the key `"g"` is removed due to its value is `None`. I realized that this was happening due to the part of the function that handles lists having an `if` rather than `elif`. As a result, cases like the one above were going through the case 1 and case 4 parts of the function, causing the original key to be reintroduced and overwriting that key in the clean dictionary.

File: agoradatatools/etl/load.py [hunk 3]
```diff
@@ -52,12 +65,15 @@ def remove_non_values(d: dict) -> dict:
...
-            for elem in value:  # value is a list
+        elif isinstance(
+            value, list
+        ):  # was missing elif before - was just if and was breaking the recursion
+            for i, elem in enumerate(value):  # value is a list
                 if isinstance(elem, dict):
-                    nested_dict = remove_non_values(elem)
-                else:
-                    cleaned_dict[key] = value
+                    value[i] = remove_non_values(elem)
```
Comment (BWMac) [inline agoradatatools/etl/load.py:73]: similar to the description of the nested dictionary case, dictionaries within lists were also not being cleaned. My proposed solution here loops through the list, runs the `remove_non_values` function on dictionaries, and then removes empty dictionaries from the list that would result from a case such as `"l": ["m", "n", {"o": None}]`

File: agoradatatools/etl/load.py [hunk 4]
```diff
@@ -96,52 +123,61 @@ def load(file_path: str, provenance: list, destination: str, syn=None):
...
-        temp_json = open(path.join(staging_path, filename), 'w+')
-        json.dump(df_as_dict, temp_json,
-                  cls=NumpyEncoder,
-                  indent=2)
+        temp_json = open(path.join(staging_path, filename), "w+")
+        json.dump(df_as_dict, temp_json, cls=NumpyEncoder, indent=2)
     except Exception as e:
         print(e)
-        temp_json.close()
+        try:  # this was failing if the `try` above fails before creating `temp_json`
```
Comment (BWMac) [inline agoradatatools/etl/load.py:158]: I noticed that if the `try` failed before the `temp_json` object was initialized, then the `temp_json.close()` line would throw an error since `temp_json` did not exist.

File: agoradatatools/etl/load.py [hunk 5]
```diff
@@ -85,7 +110,9 @@ def load(file_path: str, provenance: list, destination: str, syn=None):
         activity = Activity(used=provenance)
     except ValueError:
         print(str(provenance) + " has one or more invalid syn ids")
-        return
+        return (
+            None  # added to be more explicit and consistent with the rest of the script
```
Comment (BWMac) [inline agoradatatools/etl/load.py:114]: I changed `return` to `return None` in a few places in this function to be more explicit and to match the other functions

File: agoradatatools/etl/load.py [hunk 6]
```diff
@@ -152,21 +188,32 @@ def df_to_csv(df: pd.DataFrame, staging_path: str, filename: str):
...
-        temp_json = open(path.join(staging_path, filename), 'w+')
-        json.dump(df_as_dict, temp_json,
-                  cls=NumpyEncoder,
-                  indent=2)
+        temp_json = open(path.join(staging_path, filename), "w+")
+        json.dump(df_as_dict, temp_json, cls=NumpyEncoder, indent=2)
     except Exception as e:
         print(e)
-        temp_json.close()
+        try:  # handle case where `try` fails before temp_json is created
```
Comment (BWMac) [inline agoradatatools/etl/load.py:213]: Same as example above

File: tests/test_load.py [hunk 7]
```diff
@@ -1,33 +1,225 @@
...
+            destination="syn1111113",
+            syn=None,
+        )
+        self.patch_syn_login.assert_called_once()
+        assert "one or more invalid syn ids" in captured_output.getvalue()
+        self.patch_syn_store.assert_not_called()
+        assert test_tuple is None
+
+    def test_load_syn_store_fails_OSError(self):
+        self.patch_syn_store.side_effect = OSError  # can't induce failure in mocked object with arguments, so do manually
```
Comment (BWMac) [inline tests/test_load.py:102]: Learned how to make mocked objects throw specific errors - handy because you can't make them throw errors by passing bad arguments to them.

**Review by BWMac — COMMENTED** [review]
I know that some of this work was out of the scope for the ticket. Let me know if I misunderstood anything in the changes that I made in `load.py`. Happy to revert or revise as needed

**Comment by BWMac** [comment]
I will leave this PR as a draft until I have received feedback since it makes changes that are out of scope for this ticket.

File: agoradatatools/etl/load.py [hunk 8]
```diff
@@ -1,9 +1,12 @@
-import pandas as pd
 import json
-from os import mkdir, rmdir, path
-from . import utils
-from synapseclient import File, Activity
+from os import mkdir, path, rmdir
```
Comment (thomasyu888) [inline agoradatatools/etl/load.py]: Nit: I sort of dislike this specific way of importing `os` modules. I personally prefer doing `os.mkdir` so that I am certain the `mkdir` function comes from `os`. That said, no need to configure it.

File: agoradatatools/etl/load.py (hunk 5 above)
Comment (thomasyu888) [reply agoradatatools/etl/load.py:114]: This is great and I think follows pep8 standard when specifying a function that returns None and something else.

File: agoradatatools/etl/load.py (hunk 4 above)
Comment (thomasyu888) [reply agoradatatools/etl/load.py:158]: After looking at this some more, I think it might be more efficient to do the following.

```
try:
    df = df.replace({np.nan: None})
    df_as_dict = df.to_dict(orient="records")
    with open(path.join(staging_path, filename), "w+") as temp_json:
        json.dump(df_as_dict, temp_json, cls=NumpyEncoder, indent=2)
except Exception as e:
    print(e)
    return None
return temp_json.name
```

What do you think? I think this simplifies the code - that said, I think we can open up another ticket to improve these portions - since there are a lot of places that use `.close()`

File: agoradatatools/etl/load.py (hunk 3 above)
Comment (thomasyu888) [reply agoradatatools/etl/load.py:73]: I feel good about this function as it pertains to the scope of this ticket, but what are your thoughts on something like this: https://stackoverflow.com/questions/2544710/how-i-can-get-rid-of-none-values-in-dictionary maybe as a refactor opportunity in the future?

I do realize that what this doesn't do is remove the empty dictionaries: so this exact test case would result in: `"l": ["m", "n", {}]`

File: tests/test_load.py [hunk 9]
```diff
@@ -1,33 +1,227 @@
...
+    os.rmdir("./test_staging_dir")
+
+
+def test_delete_temp_location():
+    os.mkdir("./test_staging_dir")
+    load.delete_temp_location(staging_path="./test_staging_dir")
+    assert not os.path.exists("./test_staging_dir")
+
+
+def test_remove_non_values():
```
Comment (thomasyu888) [inline tests/test_load.py:28]: Great test!

File: tests/test_load.py (hunk 7 above)
Comment (thomasyu888) [reply tests/test_load.py:102]: Excellent!  Glad you learned how to do this!

File: tests/test_load.py [hunk 10]
```diff
@@ -1,33 +1,227 @@
...
+
+    def test_df_to_json_success(self):
+        json_name = load.df_to_json(
+            df=pd.DataFrame(), staging_path="./staging", filename="test.json"
+        )
+        self.patch_replace.assert_called_once_with({np.nan: None})
+        self.patch_to_dict.assert_called_once_with(orient="records")
+        self.patch_json_dump.assert_called_once_with(
+            self.patch_to_dict.return_value,
+            ANY,  # without mocking open() I was unable to get anything equating to `temp_json` to go here, it was failing even when the 'Expected call' and the 'Actual call' appeared to match perfectly
```
Comment (thomasyu888) [inline tests/test_load.py:152]: TIL: the `ANY` object.

**Review by thomasyu888 — CHANGES_REQUESTED** [review]
🚀 🚀 🚀   Just some comments - but overall - nice work!

File: tests/test_load.py [hunk 11]
```diff
@@ -1,33 +1,227 @@
...
+        captured_output = StringIO()
+        sys.stdout = captured_output
+        test_tuple = load.load(
+            file_path="fake/path/to/fake/file",
+            provenance=["not a syn id", "also not a syn id"],
+            destination="syn1111113",
+            syn=None,
+        )
+        self.patch_syn_login.assert_called_once()
+        assert "one or more invalid syn ids" in captured_output.getvalue()
```
Comment (thomasyu888) [inline tests/test_load.py:97]: Nice usage of StringIO() to capture stdout!

File: agoradatatools/etl/load.py (hunk 3 above)
Comment (BWMac) [reply agoradatatools/etl/load.py:73]: Do the empty dictionaries matter? I wasn't sure, but I didn't think that it would hurt to remove them

File: agoradatatools/etl/load.py (hunk 4 above)
Comment (BWMac) [reply agoradatatools/etl/load.py:158]: Yeah I agree, definitely would simplify things. Should we just open another ticket for that and any other improvements that can link to this PR?

File: agoradatatools/etl/load.py (hunk 8 above)
Comment (BWMac) [reply agoradatatools/etl/load.py]: I agree and will change it since it will be a quick change. This is the way it was before and I opted not to mess with it at the time

Comment (BWMac) [reply agoradatatools/etl/load.py]: changed in latest commit

File: agoradatatools/etl/load.py (hunk 3 above)
Comment (thomasyu888) [reply agoradatatools/etl/load.py:73]: Yes, per our discussion, my recommendation is to :

1. Run the code on the test configuration and compare older version with newer version
2. Discuss with Jess about this change if there are differences to see if it may affect downstream usage of data.

File: agoradatatools/etl/load.py (hunk 4 above)
Comment (thomasyu888) [reply agoradatatools/etl/load.py:158]: Yes, create a ticket to do this work.

Comment (BWMac) [reply agoradatatools/etl/load.py:158]: ticket created [here](https://sagebionetworks.jira.com/browse/IBCDPE-461)

File: agoradatatools/etl/load.py (hunk 3 above)
Comment (thomasyu888) [reply agoradatatools/etl/load.py:73]: Confirmed that the output does not change

**Review by thomasyu888 — APPROVED** [review]
🔥 🔥 🔥
